streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -824,27 +834,44 @@ long runOnce(final long recordsProcessedBeforeCommit) {
         }
 
         if (records != null && !records.isEmpty()) {
-            streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
+            streamsMetrics.pollTimeSensor.record(computeLatency(), now);
             addRecordsToTasks(records);
         }
 
-        if (taskManager.hasActiveRunningTasks()) {
-            final long totalProcessed = processAndMaybeCommit(recordsProcessedBeforeCommit);
-            if (totalProcessed > 0) {
-                final long processLatency = computeLatency();
-                streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessed, timerStartedMs);
-                processedBeforeCommit = adjustRecordsProcessedBeforeCommit(
-                        recordsProcessedBeforeCommit,
-                        totalProcessed,
-                        processLatency,
-                        commitTimeMs);
-            }
-        }
+        if (state == State.RUNNING) {
+            taskManager.maybeEnforceProcess();
+
+            /*
+             * Within an iteration, after N (N initialized as 1 upon start up) round of processing one-record-each on the applicable tasks, check the current time:
+             *  1. If it is time to commit, do it;
+             *  2. If it is time to punctuate, do it;
+             *  3. If elapsed time is close to consumer's max.poll.interval.ms, end the current iteration immediately.
+             *  4. If none of the the above happens, increment N.
+             *  5. If one of the above happens, half the value of N.
+             */
+            long totalProcessed;
+            long timeSinceLastPoll;
+
+            do {
+                totalProcessed = processAndMaybeCommit();
+                timeSinceLastPoll = Math.max(now - lastPollMs, 0);
+
+                if (timeSinceLastPoll / 2 >= maxPollTimeMs) {
```

Comment:
Yes, that was my plan. I'm aware that this line is basically a no-op because max.poll is Integer.MAX_VALUE, and want to do it in another PR. If people feel that we should reverse the ordering, i.e. change the default value first, then do this PR, I'm fine too.

Reply:
> should this be timeSinceLastPoll >= maxPollTimeMs / 2 ?

Gosh, my bad.

Reply:
I see. Fine with my both ways -- as long as it's intentional and we know about it, it's ok.